llama.cpp

```diff
@@ -4852,12 +4846,18 @@ static void llm_load_vocab(
     }
 
     // build token to piece caches
-    for (llama_token id = 0; id < (llama_token) n_vocab; ++id) {
-        vocab.cache_token_to_piece[id]         = llama_token_to_piece(&model, id, false);
-        vocab.cache_token_to_piece_special[id] = llama_token_to_piece(&model, id, true);
-    }
+    {
+       std::vector<llama_vocab::token> cache_token_to_piece        (n_vocab);
+       std::vector<llama_vocab::token> cache_token_to_piece_special(n_vocab);
 
-    vocab.has_cache = true;
+       for (uint32_t id = 0; id < n_vocab; ++id) {
+           cache_token_to_piece[id]         = llama_token_to_piece(&model, id, false);
+           cache_token_to_piece_special[id] = llama_token_to_piece(&model, id, true);
+       }
+
+       std::swap(vocab.cache_token_to_piece,         cache_token_to_piece);
+       std::swap(vocab.cache_token_to_piece_special, cache_token_to_piece_special);
```

Comment:
This feels good. This prevents us from getting into sticky situations if multiple threads are both trying to mess with a half-built vocab. 

As long as the cache isn't empty, then we can be pretty sure that our token is going to be present, and don't need to worry about cache misses.

There are still a few other places where we assume that the cache is going to be present, but in general this feels good I think? I need to look better and understand the vectors by which a vocab could potentially be initialized without a cache.

Reply:
Question resolved by comment at https://github.com/ggerganov/llama.cpp/pull/7587#discussion_r1619293050